### ISSUE TITLE
a11y: jump to main content doesn't work without JS

### ISSUE THREAD

**Issue by locriacyber** [facebook/docusaurus#6411]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [X] I have read the console error message carefully (if applicable).

### Description

1. "Jump to main content" has href of `#`, which isn't doing anything when Javascript isn't available.
2. Default link color too bright. Poor contrast with white background.
3. Links don't change appearance when focused. No idea which link is focused because of this. I don't need a screen reader.

### Steps to reproduce

1. Go to https://docusaurus.io/docs in Firefox Desktop
2. Disable Javascript & refresh
3. Press <kbd>Tab</kbd>, and try to use link "Skip to main content" with <kbd>Enter</kbd>

### Expected behavior

Implied in Description

### Actual behavior

Implied in Description

### Reproducible demo

https://docusaurus.io/docs

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by locriacyber** [comment]
Discovered in https://github.com/tauri-apps/tauri-docs/issues/307

**Comment by Josh-Cena** [comment]
> Default link color too bright. Poor contrast with white background.

Will be fixed by #5822

The others are valid issues. Thanks for reporting.

**Comment by locriacyber** [comment]
Only the linked color contrast is fixed by #5822 (Used the link preview https://deploy-preview-5822--docusaurus-2.netlify.app/docs/styling-layout/)

**Comment by Josh-Cena** [comment]
Indeed, my message says that only the quoted problem will be fixed😆 Maybe I didn't make it clear enough...

**Comment by Josh-Cena** [comment]
The outline is moved to JS logic in https://github.com/facebook/docusaurus/pull/3626. Apparently, the fix here would be to use `:focus-visible`, but until this point, it's not supported in Safari. (Although Safari has terrible support of tabbing anyways)

@locriacyber was wondering, why would someone needing visual aid use a website without JavaScript? I was discussing this with some other front-end developers, and seems the consensus is that JavaScript is necessary for a lot of a11y features, unless you rely on experimental WCAG features. a11y and browser compatibility—we have to make a compromise.

**Comment by slorber** [comment]
We fixed the focus outline normally @locriacyber , please let us know if it doesn't work

**Comment by locriacyber** [comment]
> why would someone needing accessibility aid use a website without JavaScript?

It's a separate issue. The browser default is to underline links when focused. This feature only needs CSS.

**Comment by Josh-Cena** [comment]
@lex111 Wonder what's your idea on this issue? Should we use programmatic focus, or should we simply use a link to an element ID? Other sites I researched (like NYT and GitHub) seem to use an anchor link. GitHub would instantly erase the hash from the location if JS is enabled.

**Comment by lex111** [comment]
Initially I didn't add explicit identifiers to try to make the STC button as workable as possible on any page. You could try to add identifiers that are resolves this issue, but why would anyone use a site with disabled JS, I still don't get it?

**Comment by lex111** [comment]
Not so sure about that, but most likely yes.

**Comment by locriacyber** [comment]
`id="main-45678987654345"` is better

**Comment by Josh-Cena** [comment]
React 18 has a `useId` hook, I think we can use that.

**Comment by mturoci** [comment]
@Josh-Cena @slorber I would like to pick up this issue, but have a few clarifying questions:

* Shall the existing JS focus be dumped entirely in favor of a simple anchor? (IMHO using native href is always preferred compared to JS so this is a good direction).
* `useId` hook cannot be used because the root docusaurus package (main package.json) depends on React 17 and theme-classic has also peer dep only 16.8 or 17. `UseId` is however only available since 18.
* I am not sure why the dynamic `useId` hook is necessary anyway. When looking at the documentation, it seems like the id shall be added to Layout > mainWrapper div and since I assume every page will always have at most 1 Layout, then static ID shall be enough. Please correct me if I am missing something.

**Comment by slorber** [comment]
Something to consider:
- our current skipToContent behavior focuses in priority an existing `<main>` tag over 
- we are keeping this behavior (for now)
- we are introducing an explicit "id" that is used more as a "fallback" in case no `<main>` tag exists. In this case, focusing on the `<Layout>` children container is reasonable behavior.
- when JS is disabled, it's not possible to replicate the same behavior (`<main>` used in priority). Afaik it's not possible to have a link target a `<main>` element that may not have an id

For some pages, like the docs, it leads to slightly different behaviors with/without JS.

Test page: https://deploy-preview-8204--docusaurus-2.netlify.app/docs/


With JS: the document markdown content gets focused on first tab (`<main>`):

<img width="1912" alt="CleanShot 2022-10-19 at 16 44 32@2x" src="https://user-images.githubusercontent.com/749374/196724052-99dfad49-3212-476a-a122-1376b0da6bea.png">


Without JS: the docs sidebar get focused on first tab (layout children, fallback behavior)

<img width="1912" alt="CleanShot 2022-10-19 at 16 43 29@2x" src="https://user-images.githubusercontent.com/749374/196723808-b2166427-c84e-492e-bf9c-a79bf2739409.png">

---

That is likely not a big deal. I'd rather avoid changing everything of this implementation for now, and preserve the "focus main in priority" behavior for retro compatibility.
